[SPEC] Add relative paths to v4 spec (#15630)
apache/iceberg: format/spec.md

The full set of changes are listed in [Appendix E](#version-3).

### Version 4: Metadata Structure, Statistics Representation, Relative Paths, Expressions

#### Relative Paths

Version 4 of the Iceberg spec adds support for relative paths in metadata, enabling tables to be relocated without rewriting metadata files. Key changes include:

> * Support for relative paths in all metadata tracked path fields, resolved against the table's base location

Author: I'll update and scrub the rest of the updates to reference `location`

> * Support for relative paths in all metadata tracked path fields, resolved against the table's base location

Author: I've updated this, but left a number of references to "path" where that's the canonical term (e.g. Path Resolution, Path Relativization, etc)

Tables do not require rename, except for tables that use atomic rename to implement the commit operation for new metadata files.

> ### Relative Locations in Metadata

Reviewer: Is `File Paths in Metadata` slightly better for a summary sub section? It also matches the later section of `Paths in Metadata`

We can also briefly talk about in V3 and prio, all file locations use absolute path.

Author: That's what it was originally and @rdblue asked to change it to locations.

Reviewer: path or location is fine either way.

I was mainly saying if we should remove the `Relative` in the section title. Just `File Locations in Metadata`

> Version 4 of the Iceberg spec adds support for relative locations in metadata, enabling tables to be relocated without rewriting metadata files. Key changes include:

Reviewer 2: Minor: Since this covers what v4 does, it would be nice to also say that in v3 all paths were fully-qualified.

Author: This is going to get a little pedantic, but I think that's fair to include and have the following comment for reference.  

The issue is that there is disagreement about what 'absolute' and 'fully-qualified' actually mean.  In many contexts, they're considered equivalent.  However, in other contexts, 'absolute' would not require a scheme, while 'fully-qualified' does require a scheme.  Though, at the same time, 'fully-qualified' also allows for relative references.

I think we want to define 'absolute' as requiring a scheme without relative or special references (as we have done here).  That allows to at least make the argument that 'fully-qualified' is what the spec previously expected (though only explicitly called out for the `referenced_data_file` field).  We were not explicit in the spec about the contents of other fields.

* Support for relative locations in all metadata tracked path fields, resolved against the table's base location
* The table `location` field becomes optional, allowing the table location to be:
  * Provided by an owning catalog
  * Inferred from the metadata file location or storage layout

>   * Supplied directly where necessary

Reviewer: what does `supplied directly` mean? the `location` field from table metadata? if yes, might be more clear to spell it out explicitly.

Author: No, this would be for a path based access to the table.  So if you have something like duckdb:

```
SELECT *
FROM iceberg_scan('s3://your-bucket/path/to/iceberg/table');
```

The location of the table must be provided by that implementation.  Either you can have a catalog provide it, you can infer from the `metadata.json` location, or you can directly supply it (the above isn't a great example since you need to know the specific metadata.json version which is not included).

Author: A better example might be for `register_table` where currently you must only provide the path to the `metadata.json` file.  In v4, if the location field is omitted and you want to register a table, you'll need to supply both the `metadata.json` and independently the `table-location` to the register call.

This is an example of something we'll have to update both in the REST spec and the register procedures.

Reviewer: I see. The first 2 bullet points are for read path on how to resolve the base location. This bullet is for the write path (like `register_table` or `create_table`). But the current wording `Supplied directly where necessary` seems a bit too abstract to get the right context. 

what about `Supplied directly when creating or registering a table`?

Author: It's not specific to just those two operations though.  Anywhere/anything that loads directly must know the `metadata.json` to load and the table location.  So we shouldn't constrain it to specific use cases in the spec.

Reviewer: fair enough

> * Formal definitions for path types, path resolution, and path relativization

Reviewer 2: These are great high-level changes for Appendix E, but they aren't very helpful in the overview. In the overview.

For the overview, I'd say:

> Iceberg supports relative locations in metadata in v4 and later. In earlier versions, locations must be fully-qualified.
>
> Relative locations are allowed in all metadata tracked path fields and are resolved against the table's base location. The table's location may be fixed in table metadata or inferred, but is intended to be managed and supplied by a catalog.
>
> Requirements for relativization and resolution are in [Relative Paths](...)

The full set of changes are listed in [Appendix E](#version-4).

## Specification

> ### Terms

Reviewer 2: I'm good making tweaks liek this, but do we want to include them in this PR? Usually we want to review these separately because this affects the auto-generated table of contents.

Author: This actually fixes a minor rendering issue with the table of contents, but doesn't affect the indentation or structure.  Because it fixes the rendering for the next few sections (which is material to this PR) I'd just include it here.  If you really think there's substance to the change, I can split it out.

Tables do not require rename, except for tables that use atomic rename to implement the commit operation for new metadata files.

### File Locations in Metadata

> All location fields in format versions 3 and prior contain fully-qualified paths.

Reviewer 3: ```suggestion
All location fields in format version 3 and prior contain fully-qualified paths.
```

Author: In this sentence 'versions' refers to multiple versions (3 and prior) so the plural form is correct.

Reviewer 4: It might be clearer in first glance with `in format versions prior to version 4`.

> Version 4 of the Iceberg spec adds support for relative locations in metadata, enabling tables to be relocated without rewriting metadata files. Relative locations are allowed in all metadata tracked location fields and are resolved against the table's base location. The table's location may be fixed in table metadata or inferred, but is intended to be managed and supplied by a catalog. Requirements for relativization and resolution are in [Relative Paths](#path-resolution)

Reviewer 4: nit: missing `.` at the end.

Reviewer 4: do you want to link to `#paths-in-metadata`?

Reviewer 3: +1, that includes Path Relativization and Path Resolution. `Relative Paths` is a little confusing if it does not link to the section with the same name.

Tables do not require rename, except for tables that use atomic rename to implement the commit operation for new metadata files.

### File Locations in Metadata

All location fields in format versions 3 and prior contain fully-qualified paths.

Version 4 of the Iceberg spec adds support for relative locations in metadata, enabling tables to be relocated without rewriting metadata files. Relative locations are allowed in all metadata tracked location fields and are resolved against the table's base location. The table's location may be fixed in table metadata or inferred, but is intended to be managed and supplied by a catalog.

> Requirements for relativization and resolution are in [Relative Paths](#path-resolution)

Reviewer 2: If you intended for this to be a separate paragraph, I think you have to add an empty newline.

Author: No, the wrap was unintentional.  Fixed.

### Paths in Metadata

Path strings stored in Iceberg metadata files are classified as one of two types:

* **Absolute path** -- A path string that includes a [URI scheme](https://datatracker.ietf.org/doc/html/rfc3986#section-3.1) (e.g., `s3://`, `gs://`, `hdfs://`, `file:///`). Absolute paths are used as-is without modification.
* **Relative path** -- A path string that does not include a URI scheme. Relative paths must be resolved against the table's base location before use.

Prior to v4, all path fields must contain absolute paths. Starting with v4, path fields may contain either absolute or relative paths. Directory navigation symbols (`.` and `..`) and other file system conventions are not supported in relative paths.

#### Path Resolution

Path resolution is the process of producing an absolute path from a relative path by combining it with the table's base location. If a path is absolute, it is used as-is. If a path is relative, it is concatenated with the table location to produce an absolute path:

* If the path contains a URI scheme, it is absolute and is used without modification.

> * If the path does not contain a URI scheme, the resolved path is the table location followed by the relative path.

Author: Based on our definition, it is not a valid local abosolute path and I think that is a fair assertion.  There are many implementations that require a `file:///` or `local:///` prefix for local addressing.

For us to interpret that as a local path, we would also need to interpret the path separator, which we do not want to define.  I think the right path forward is to use the definition of absolute as stated and require full URI for aboslute paths.

Paths used as prefixes must not end in a path separator. The relative portion is appended to the prefix without introduction of any additional separator characters.

#### Path Relativization

Path relativization is the process of converting an absolute path to a relative path by removing the table location prefix. This is used when persisting paths to metadata files.

> * If an absolute path starts with the table location, the table location prefix is removed and the remaining relative portion is stored.

Reviewer 2: Can be removed or must be removed? I think it is optional to relativize paths because older versions do not and those are still valid. We can state that paths should be relativized in v4 and later, but I don't want to imply that they must be.

Tables do not require rename, except for tables that use atomic rename to implement the commit operation for new metadata files.

### Paths in Metadata

Path strings stored in Iceberg metadata files are classified as one of two types:

* **Absolute path** -- A path string that includes a [URI scheme](https://datatracker.ietf.org/doc/html/rfc3986#section-3.1) (e.g., `s3://`, `gs://`, `hdfs://`, `file:///`). Absolute paths are used as-is without modification.
* **Relative path** -- A path string that does not include a URI scheme. Relative paths must be resolved against the table's base location before use.

> Prior to v4, all path fields must contain absolute paths. Starting with v4, path fields may contain either absolute or relative paths. Directory navigation symbols (`.` and `..`) and other file system conventions are not supported in relative paths.

Reviewer 2: Normalization is not supported across storage systems. For instance, in S3, `s3://bucket/a/b//c` is distinct from `s3://bucket/a/b/c`. We do not want to get involved with rewriting or normalizing paths because that causes issues: Iceberg can have two paths that are not equal but would normalize to the same thing.

This is why Iceberg takes a hands-off approach: `FileIO` implementations are responsible for these weird cases. If paths are not supported (`b//c` or `b/../c` or `b/./c`) then the `FileIO` should reject them. Other `FileIO` implementations can (and should) support those paths since they are valid -- but should not support any special interpretation.

> Prior to v4, all path fields must contain absolute paths. Starting with v4, path fields may contain either absolute or relative paths. Directory navigation symbols (`.` and `..`) and other file system conventions are not supported in relative paths.

Author: @mxm Unfortunately, I don't think we can require that implementations reject these paths.  They're already supported by some of the underlying implementations, so adding a restriction may not be compatible. 

As @rdblue pointed out above, the sequences are actually valid for some storage implementations.  For example, S3 will allow a key with `..`, but it doesn't treat it as navigation, but rather as part of the key sequence.

The intent here is to not support usage of character sequences for the purposes of file system handling or navigation that some file systems support.  It's a little hard to explicitly enforce this.

> Prior to v4, all path fields must contain absolute paths. Starting with v4, path fields may contain either absolute or relative paths. Directory navigation symbols (`.` and `..`) and other file system conventions are not supported in relative paths.

Reviewer 5: @danielcweeks A couple of points to clarify:

1. Absolute paths today can already contain file system directives. 
2. If we want to state in the spec that file system directives should not be used, we might want to state that for both relative and absolute paths. Currently, we are just making that statement for relative paths.
3. I'm not suggesting to reject file system directives; we can't, the table metadata merely stores strings.
4. It is up to the FileIO implementation to reject directives (or not).
  a. In the case of S3, it would be wrong to reject ".." or similar because they are legitimate path elements.
  b. In the case of other file systems, they are free to decide.

I tried to convey that in the suggested change above. Refining here:

```suggestion
Prior to v4, all path fields must contain absolute paths. Starting with v4, path fields may contain either absolute or relative paths. Paths (relative or absolute) must not include symbols which the underlying file system interprets as directives (e.g. `..` in file systems which treat this as the parent directory). It is up to the file system implementation to validate these paths.
```

Does that make sense?

> Prior to v4, all path fields must contain absolute paths. Starting with v4, path fields may contain either absolute or relative paths. Directory navigation symbols (`.` and `..`) and other file system conventions are not supported in relative paths.

Author: @mxm I think we're getting to the point that we're over-clarifying and creating carveouts for implementations.  I think we should keep this simple and not worry about what underlying implements must or must not do.  

To be clear, v4 doesn't introduce this as a new problem (it exists for all versions) and we haven't needed to clarify this (with the exception of some accidental implementation gaffes like `//` in s3).

If this is a point of contention, maybe we should remove it.

> Prior to v4, all path fields must contain absolute paths. Starting with v4, path fields may contain either absolute or relative paths. Directory navigation symbols (`.` and `..`) and other file system conventions are not supported in relative paths.

Reviewer 5: Fair. This issue is not specific to V4. Let's keep it simple and drop the mention of file directives.

#### Path Resolution

Path resolution is the process of producing an absolute path from a relative path by combining it with the table's base location. If a path is absolute, it is used as-is. If a path is relative, it is concatenated with the table location to produce an absolute path:

* If the path contains a URI scheme, it is absolute and is used without modification.
* If the path does not contain a URI scheme, the resolved path is the table location followed by the relative path.

> Paths used as prefixes must not end in a path separator. The relative portion is appended to the prefix without introduction of any additional separator characters.

Reviewer 6: maybe add "if . or .. is found as a path element in an relative or absolute path then the resolution must fail"

Author: These are valid sequences, though, so I don't think we can require failing.  I could see some implementations that actually use them for navigation (like HadoopFileIO) rejecting them, but even then it's a little inconsistent.  I'm not even sure how this would work since `Path` will treat it as navigation even if the path with `..` exists as a key in S3.

Regardless, I think we need to support them as characters, but not as navigation.

* **Manifest** -- A file that lists data or delete files; a subset of a snapshot.
* **Data file** -- A file that contains rows of a table.
* **Delete file** -- A file that encodes rows of a table that are deleted by position or data values.
* **Absolute path** -- A path string that includes a [URI](https://datatracker.ietf.org/doc/html/rfc3986#section-3.1) scheme and can be used directly.
* **Relative path** -- A path string without a URI scheme that must be [resolved](#path-resolution) against the table location.

> 

Reviewer: The spec defines **Absolute path** and **Relative path** in the bullets above, but **fully-qualified path** is used here (and on lines 136, 206, 1741, and as a row label in the example table at line 214) without a definition.

The term is doing work that the two defined terms can't quite cover. Line 206 says "Any path from a manifest produced prior to v4 is a fully-qualified path and must be produced with a URI scheme **if the scheme was omitted** to be consistent with V4 paths." That implies a fully-qualified path may *not* have a URI scheme — which conflicts with the v4 **Absolute path** definition (which requires one). The "Missing scheme" example row at line 215 reinforces this: a pre-v4 path can lack a URI scheme yet still be treated as a usable, non-relative path.

Two cleanup options:

1. **Define the term**, e.g.:
   > **Fully-qualified path** -- A path used in v3 and earlier metadata that is treated as ready-to-use without resolution against the table location. May contain a URI scheme; when omitted, an implementation-defined default scheme is prepended on read.

2. **Drop the term** and use **Absolute path** everywhere, with a clarifying clause that pre-v4 paths omitting a URI scheme are treated as absolute after the default scheme is prepended.

Option 2 feels cleaner — one fewer term, and the migration rule on lines 1741-1742 reads more naturally as "v3 paths are absolute paths (with the scheme prepended if missing)."

Author: I disagree with this (I'm not sure if this was AI generated or your opinion).  Versions prior to V4 were defined in the spec already as either "URI with scheme" or "fully-qualified".  Those were the existing terms in the spec.

I don't think we should go back to further define those terms as we may introduce new requirements on older versions.  The new terms apply to V4 and the behaviors being introduced in this revision.

You need to take into consideration backward compatibility, which means that we cannot apply option 2 as it would redefine prior versions of the spec.

Reviewer: In the end, every comment reflects my opinion because I carefully reviewed, edited, or dropped the comments before publishing.

I just checked the current table spec. There is only one mentioning of fully qualified for the data file fields, where the wording matching the definition of absolute path.
```
143 referenced_data_file | string | Fully qualified location (URI with FS scheme) of a data file that all deletes reference
```

I was saying that it was not explicitly defined in the terms list. does it make sense to define the term of fully qualified path in the bullet list formally? or we can clarify what absolute path can behavior slightly differently in v3 and prior.

Author: The definition for a `file_path` is "Full URI for the file with FS scheme", which I believe was intended to be the same as "fully-qualified" (though it's unclear because it wasn't defined).

I don't think we should go back and try to define terms we didn't define originally since it's ambiguous and clarifying at this point could cause more confusion.  I think it's better to focus on defining what v4 requires moving forward.

Reviewer 2: I agree with Dan here. We don't want to introduce more definitions for old versions. We should carry them forward and let them be interpreted as they were before.

The only place where this matters is the scheme, which was specifically called out as required. It was never optional, but we know that paths were sometimes created without it. The note about supplying a scheme for v3 and earlier isn't changing the requirement, it is stating that if you produce paths without a scheme when reading, it's going to break in v4 because we are more strict and would decide to handle paths without a scheme as relative.

Reviewer: fair enough

Writers are not allowed to commit files with a partition spec that contains a field with an unknown transform.

### Paths in Metadata

> Path strings stored in Iceberg metadata location fields are classified as one of two types:

Reviewer 7: Nit: There are a few references to "fully qualified path" later in the context of v3 and prior, without it being explicitly defined. Since we're classifying paths into two types
   below, it might be worth briefly noting that fully qualified paths from v3 and prior are considered absolute paths. This could help connect the dots more easily.

Author: We don't want to do this (see other comments on this topic).  We don't want to go back and define things that weren't defined for prior versions since it could introduce additional requirements on older versions.  The prior spec only referred to "fully-qualified" and "URI with Scheme" for fields and we're not trying to rewrite those versions of the spec.

Reviewer 7: Thanks Dan, found it in https://github.com/apache/iceberg/pull/15630#discussion_r3250983873

* **Absolute path** -- A path string that starts with a [URI scheme](https://datatracker.ietf.org/doc/html/rfc3986#section-3.1) (e.g., `s3:`, `gs:`, `hdfs:`, `file:`). Absolute paths are used as-is without modification.
* **Relative path** -- A path string that does not start with a URI scheme. Relative paths must be resolved against the table's base location before use.

> Prior to v4, all path fields must contain fully-qualified paths. Starting with v4, path fields may contain either absolute or relative paths. [Relative resolution within a URI](https://datatracker.ietf.org/doc/html/rfc3986#section-5.2) (e.g. `.` and `..`) and other file system navigation conventions are not supported in relative paths.

Reviewer 8: ```suggestion
Prior to v4, all path fields must contain absolute paths. Starting with v4, path fields may contain either absolute or relative paths. [Relative resolution within a URI](https://datatracker.ietf.org/doc/html/rfc3986#section-5.2) (e.g. `.` and `..`) and other file system navigation conventions are not supported in relative paths.
```

Author: There are a number of similar comments about this distinction, but `fully-qualified` (or URI with scheme) is actually the terminology that was used for v1-v3.  V4 is distinctly different and part of this change is the shift in terminology, which is why we're not attempting to redefine the usage in prior versions.

#### Path Resolution

Path resolution is the process of producing an absolute path from a relative path by combining it with the table's base location:

> * If the path contains a URI scheme, it is absolute and is used without modification.

Reviewer 9: nit: this says "If the path contains a URI scheme" while L191 uses "starts with a URI scheme".

i think we should probably align and use "starts with" here to be more explicit

L191:
> * **Absolute path** -- A path string that starts with a [URI scheme](https://datatracker.ietf.org/doc/html/rfc3986#section-3.1) (e.g., `s3:`, `gs:`, `hdfs:`, `file:`). Absolute paths are used as-is without modification.

Author: Updated wording.  Technically, they're the same since a URI scheme is defined by the URI spec (it can't be anything else).  However, I'll change to starts with for consistency.

* If the path does not contain a URI scheme, the resolved path is the table location followed by the relative path joined by the URI separator character `/`.

> The relative portion is joined to the prefix (table location) without consideration of any additional separator characters. The recommended convention for table location is to not end in a path separator because the join process would add a second separator character. (See example below).

Reviewer 10: The part "without consideration of any additional separator characters" feels a bit contradictory to the previous sentence that says the join is via `/`. Is this a leftover from the previous iteration?

Author: No, it means that if there are other separators included in the base location or the relative part, those will not be removed or deduplicated.

> The relative portion is joined to the prefix (table location) without consideration of any additional separator characters. The recommended convention for table location is to not end in a path separator because the join process would add a second separator character. (See example below).

Reviewer 4: nit: `character. (See example below.)` or `character (see example below).`

> The relative portion is joined to the prefix (table location) without consideration of any additional separator characters. The recommended convention for table location is to not end in a path separator because the join process would add a second separator character. (See example below).

Reviewer 11: I'm not sure we need the examples for duplicate separator. I think that's pretty straight forward?

Author: Others asked for this explicitly to show what is expected if you sufix/prefix with a separator and what the behavior would look like.  The point is to show that you do not de-dup or strip them.

> Paths in manifests produced prior to v4 are fully-qualified and must be produced with a URI scheme if the scheme was omitted to be consistent with V4 paths.

Reviewer 12: nit: I had to read this sentence multiple times to understand it

Reviewer 4: Me too ;), might need a comma somewhere.

> Paths in manifests produced prior to v4 are fully-qualified and must be produced with a URI scheme if the scheme was omitted to be consistent with V4 paths.

Reviewer 4: It looks we use `v4` instead of `V4` in all other places.

Examples of path resolution:

|                     | Format Version | Table Location        | File Path                                 | Resolved Path                              | Description                         |
|---------------------|----------------|-----------------------|-------------------------------------------|--------------------------------------------|-------------------------------------|
| Relative Path       | v4             | s3://bucket/db/table  | data/00000-0.parquet                      | s3://bucket/db/table/data/00000-0.parquet  | Path parts are joined on `/`        |

> | Absolute Path       | v4             | s3://bucket/db/table  | hdfs:/wh/db/table/data/00000-0.parquet    | hdfs://wh/db/table/data/00000-0.parquet    | Absolute path is used               |

Reviewer 10: This doesn't look right. There is an extra `/` in the protocol part of the resolved path. It would give the impression that we are doing some path normalization.

Author: Nice catch, I think that was a typo.  Will fix.

> | Duplicate separator | v4             | s3://bucket/db/table/ | data/00000-0.parquet                      | s3://bucket/db/table//data/00000-0.parquet | Join results in duplicate `//`      |

Reviewer 10: Thanks for calling these out explicitly.

| Duplicate separator | v4             | s3://bucket/db/table  | /data/00000-0.parquet                     | s3://bucket/db/table//data/00000-0.parquet | Join results in duplicate `//`      |
| Fully-qualified     | v3 and earlier | s3://bucket/db/table  | s3://bucket/db/table/data/00000-0.parquet | s3://bucket/db/table/data/00000-0.parquet  | Fully-qualified path is used        |
| Missing scheme      | v3 and earlier | /wh/db/table          | /wh/db/table/data/00000-0.parquet         | hdfs://wh/db/table/data/00000-0.parquet    | Scheme is prepended for consistency |

#### Path Relativization

Path relativization is the process of converting an absolute path to a relative path by removing the table location prefix. This is used when persisting paths to metadata files.

* If an absolute path starts with the table location immediately followed by a separator character, the relative path is the remainder of the string after the separator character.

> * If an absolute path does not start with the table location immediately followed by the separator character, it is stored as an absolute path.

Reviewer 7: nit: It might be helpful to explicitly highlight `stored as an absolute path without modification`

Author: I'm not sure we need to clarify.  I think that's important to say for consuming from metadata, but how the persisted path is arrived at is different.  If something is producing the path, it can pretty much do what ever it wants with the structure as long as it's absolute when it's first persisted.

This is a little nuanced, but I think it would be overreaching in this particular context.

#### Table Location Specification

> When the `location` field is present in table metadata, it is used directly as the table's base location. When the `location` field is not present (v4 and later), the table location must be provided.  How the table location is persisted or determined when not specified in metadata is not a table-level concern; catalogs should provide a table's location

Reviewer 4: nit: missing `.` at the end.

> When the `location` field is present in table metadata, it is used directly as the table's base location. When the `location` field is not present (v4 and later), the table location must be provided.  How the table location is persisted or determined when not specified in metadata is not a table-level concern; catalogs should provide a table's location

Reviewer 11: ```suggestion
When the `location` field is present in table metadata, it is used directly as the table's base location. When the `location` field is not present (v4 and later), the table location must be maintained and provided by the catalog. ```

Author: We don't want to restrict this to catalogs only.

Please see this comment: https://github.com/apache/iceberg/pull/15630#discussion_r3251040825

Writers are not allowed to commit files with a partition spec that contains a field with an unknown transform.

### Paths in Metadata

Path strings stored in Iceberg metadata files are classified as one of two types:

* **Absolute path** -- A path string that includes a [URI scheme](https://datatracker.ietf.org/doc/html/rfc3986#section-3.1) (e.g., `s3:`, `gs:`, `hdfs:`, `file:`). Absolute paths are used as-is without modification.
* **Relative path** -- A path string that does not include a URI scheme. Relative paths must be resolved against the table's base location before use.

Prior to v4, all path fields must contain absolute paths. Starting with v4, path fields may contain either absolute or relative paths. Directory navigation symbols (`.` and `..`) and other file system conventions are not supported in relative paths.

#### Path Resolution

Path resolution is the process of producing an absolute path from a relative path by combining it with the table's base location:

* If the path contains a URI scheme, it is absolute and is used without modification.

> * If the path does not contain a URI scheme, the resolved path is the table location followed by the relative path.

Reviewer 10: The line clearly defines resolution as string concatenation, not RFC 3986 reference resolution.  The RFC has defined the reference resolution algorithm [here](https://datatracker.ietf.org/doc/html/rfc3986#section-5.2.2).  

This sentence is easy to miss and an implementer could reasonably reach for their language's URL resolver (like Rust's `Url::join()`) and get wrong results for `/` prefixed paths.

Specifically Rust `Url::join("s3://bucket/table/", "/metadata/foo.parquet")` hits the `if (R.path starts-with "/") then[...]`  branch in the RFC and incorrectly produces `s3://bucket/metadata/foo.parquet`.

Should we add a warning somewhere in the Path Resolution section?

Author: I feel like this line and the line below clearly state how this is handled.  I'm not sure how much we want to focus on how it can be done wrong in the spec and the reference implementation does a lot of trailing slash removal for table locations.

If someone read the next line, it's clear that this would be wrong:

```
Paths used as prefixes must not end in a path separator . . . 
```

Author: I also added an chart of examples below

Writers are not allowed to commit files with a partition spec that contains a field with an unknown transform.

### Paths in Metadata

Path strings stored in Iceberg metadata location fields are classified as one of two types:

* **Absolute path** -- A path string that includes a [URI scheme](https://datatracker.ietf.org/doc/html/rfc3986#section-3.1) (e.g., `s3:`, `gs:`, `hdfs:`, `file:`). Absolute paths are used as-is without modification.

> * **Relative path** -- A path string that does not include a URI scheme. Relative paths must be resolved against the table's base location before use.

Reviewer 2: Nit: "start with" instead of "include", although I think it is likely clear.

> Prior to v4, all path fields must contain fully-qualified paths. Starting with v4, path fields may contain either absolute or relative paths. [Relative resolution within a URI](https://datatracker.ietf.org/doc/html/rfc3986#section-5.2) (e.g. `.` and `..`) and other file system navigation conventions are not supported in relative paths.

Reviewer 2: This seems to imply that resolution was previously allowed, which isn't the case. I like the older wording that things like `.` and `..` have no special meaning.

I would bring back the older version and then add the link to the RFC as a clarification: The relative resolution components defined by the RFC have no special meaning and are opaque to Iceberg.

Author: We never explicitly prohibited it and I don't feel like we should be adding requirements that weren't there previously.

I didn't change the wording, so I'm not sure what you're referring to. I added the reference to the URI that describes these behaviors since other people were confused about what kind path evaluations we were referring to.

Reviewer 2: Yeah, I guess you're right. I thought sure it was changed.

I don't like saying that these aren't supported in relative paths, specifically. If you want to leave the possibility open in v3 and earlier, we can do that. But I think we should not leave open the possibility that absolute paths can contain directory navigation or other resolution patterns. I would just say that absolute and relative paths do not support these things. How about a separate paragraph:

> [Relative resolution within a URI](https://datatracker.ietf.org/doc/html/rfc3986#section-5.2) (e.g. `.` and `..`) and other file system navigation conventions are not supported in absolute or relative paths.

That way it is clear you can't use these conventions in either one. And making it a separate paragraph avoids implying that they are okay for v3 paths. We didn't specify they are not allowed, but we also chose not to implement any interpretation when issues came up, like whether `//` should be normalized or `./` should be removed.

Reviewer 11: We never prohibited this previously but it was basically FileIO dependent on whether these characters would have any kind actual meaning. We actually do follow one set of rules here, we never make a // in our own path building within the Java API. We have a bunch of "strip trailing /" code to prevent someone from using S3FileIO and HadoopFileIO from getting different results. ("bar//baz" resolving to "bar/baz" in S3A and "bar//baz" in S3FileIO)

I agree with @rdblue here that it's probably a good note to have in the spec that paths are treated as pure strings. No POSIX resolution should be assumed. I'm also fine with adding a note about absolute paths as well since I don't think we ever had this well defined but anyone using posix style things in their paths is probably making a mistake...

#### Path Resolution

Path resolution is the process of producing an absolute path from a relative path by combining it with the table's base location:

* If the path contains a URI scheme, it is absolute and is used without modification.
* If the path does not contain a URI scheme, the resolved path is the table location followed by the relative path joined by the URI separator character `/`.

> Paths used as prefixes should not end in a path separator. The relative portion is joined to the prefix without consideration of any additional separator characters.

Reviewer: Two issues with this paragraph after the latest edit:

1. **`must` → `should` introduces a real ambiguity.** With the new join-with-`/` rule on line 202, a permitted table location of `s3://bucket/table/` plus a relative `data/file` produces `s3://bucket/table//data/file` (double slash). The spec doesn't say whether implementations must reject, normalize, or accept that. Either keep `must`, or add a normalization clause covering trailing-separator table locations.

2. **"without consideration of any additional separator characters"** is a bit unclear. The new wording can be read as "any separators already present are ignored," which contradicts line 202 (the join *adds* a `/`). Suggest rewriting to: "no additional separator characters are introduced beyond the single `/` join character."

Author: I don't agree with these comments and it's not consistent with what we've discussed.  We're trying to treat the location fields as opaque as possible and still support the maximum number of scenarios.  The `should` requirement here is intentional in that we still want a way to represent a `//` in a path since it is technically valid in some storage solutions.

The "without consideration" is also intentional in that the strings should not be interpreted to try to normalize the path values.

> Paths used as prefixes should not end in a path separator. The relative portion is joined to the prefix without consideration of any additional separator characters.

Reviewer 2: I would use primarily the second sentence because it is clearly defining what happens. The first sentence is more of a recommendation to provide context:

> The relative portion is joined to the prefix (table location) without consideration of any additional separator characters. The recommended convention for table location is to not end in a path separator because the join process would add a second separator character. (See example below).

> Any path from a manifest produced prior to v4 is a fully-qualified path and must be produced with a URI scheme if the scheme was omitted to be consistent with V4 paths.

Reviewer: this sentence reads a bit awkward. is this more clear?

"Paths in pre-v4 manifests are fully-qualified. When a pre-v4 path omits a URI scheme, readers must prepend a scheme to produce a v4-consistent absolute path."

Author: I cleaned this up a little.

Examples of path resolution:

|                 | Format Version | Table Location       | File Path                                | Resolved Path                             | Description |
|-----------------|----------------|----------------------|------------------------------------------|-------------------------------------------|-------------|
| Relative Path   | v4             | s3://bucket/db/table | data/00000-0.parquet                     | s3://bucket/db/table/data/00000-0.parquet | Path parts are joined on `/` |

> | Absolute Path   | v4             | s3://bucket/db/table | hdfs:/wh/db/table/data/00000-0.parquet   | hdfs://wh/db/table/data/00000-0.parquet   | Absolute path is used |

Reviewer 2: Should we add examples to show that trailing `/` and starting `/` cause separator duplication?

Author: I'll add some, but we're not explicitly prohibiting it, but rather it would produce a undesirable result.

Reviewer 2: Yes, exactly. What we want is to make sure people understand that not removing the separator will cause a double separator. There is no normalization logic allowed or required to catch and prevent it.

| Fully-qualified | v3 and earlier | s3://bucket/db/table | s3://bucket/db/table/data/00000-0.parquet | s3://bucket/db/table/data/00000-0.parquet | Fully-qualified path is used |

> | Missing scheme  | v3 and earlier | /wh/db/table         | /wh/db/table/data/00000-0.parquet        | hdfs:/wh/db/table/data/00000-0.parquet    | Scheme is prepended for consistency |

Reviewer 2: Should this use `file:` or `hdfs:`? I'd lean toward `file:` because I think that was more common. But I could be wrong. I just haven't seen many paths that are `//namenode:8020/wh/db/table`.

Author: No, `file:` is not more common.  The typical configuration for hdfs clusters will include the default fs config in the `core-site.xml`.  I don't know of any practical situations where people are migrating from local file systems.  

The example you give would typically be encoded in the cluster configuration like:

```
<configuration>
  <property>
    <name>fs.defaultFS</name>
    <value>hdfs://namenode-host:8020</value>
    <description>The default filesystem URI</description>
  </property>
</configuration>
```

References to files would just be encoded as: `/wh/db/table`

The default fs will be handled by the hdfs FileSystem implementation.  The real use case we're covering is these types of deployments in hdfs.

Reviewer 2: I thought that would end up producing `hdfs:/wh/db/table` and not `/wh/db/table`. It's been a while since I've had to worry about this though!

#### Path Relativization

Path relativization is the process of converting an absolute path to a relative path by removing the table location prefix. This is used when persisting paths to metadata files.

> * If an absolute path starts with the table location, the table location prefix should be removed along with the separator character and the remaining relative portion stored.

Reviewer: The new wording — "removed along with the separator character" — is the closest the spec gets to closing the prefix-collision case discussed in apache/iceberg#16174 (https://github.com/apache/iceberg/pull/16174#discussion_r3228742112), but it's only implicit. There can be two different interpretation of this text:

- **Strict**: the rule applies only when the prefix is followed by the separator. If the next character isn't `/`, the prefix-removal-with-separator can't be performed, so the absolute path is stored.
- **Lax**: the rule says "strip the prefix, then strip the separator if present." `relativize("s3://bucket/table", "s3://bucket/table_v2/file")` would still strip the prefix and produce `_v2/file`.

The lax reading doesn't close the prefix-collision case discussed in apache/iceberg#16174 (https://github.com/apache/iceberg/pull/16174#discussion_r3228742112). Worth pinning this down explicitly. Suggested wording:

> An absolute path is considered under the table location if and only if it starts with the table location followed by the URI separator `/`. In that case, the table location and the separator are removed, and the remaining portion is stored as a relative path. Otherwise, the absolute path is stored.

Also worth adding one more row to the resolution example table (around line 212) that pins this case — e.g. a sibling file at `s3://bucket/db/table_v2/file.parquet` against `s3://bucket/db/table` showing the absolute path is stored, not relativized.

Author: I've updated the wording per @rdblue recommendation and this should make this explicitly clear.

> * If an absolute path starts with the table location, the table location prefix should be removed along with the separator character and the remaining relative portion stored.

Reviewer 2: I think we have to be more strict. If a path starts with the table location immediately followed by a separator character, the relative path is the remainder of the string after the separator character.

I want to avoid ambiguity in the case of a table location that is a prefix to the middle of a name, like `s3://bucket/db/tab` and `s3://bucket/db/table`. If the separator character isn't immediately after the `/`, then relativization can't happen.

> * If an absolute path does not start with the table location, it is stored as an absolute path.

Reviewer 2: Table location immediately followed by the separator character

#### Table Location Specification

> When the `location` field is present in table metadata, it is used directly as the table's base location. When the `location` field is not present (v4 and later), the table location must be provided.  How the table location is persisted or determined when not specified in metadata is not a table-level concern; catalogs should and provide a table's location

Reviewer: Two issues in this paragraph:

1. **Grammatical error**: "catalogs should and provide a table's location" is missing a verb (e.g. "manage and provide"), and the sentence has no closing period.
2. **Implicit actor**: "the table location must be provided" doesn't say by whom, then the next clause silently assumes the catalog.

Suggested rewrite:

> When the `location` field is present in table metadata, it is used directly as the table's base location. When the `location` field is not present (v4 and later), the catalog must provide the table location. How the catalog persists or determines a table's location is outside the scope of this spec.

Changes:

- Names the catalog as the actor in the second sentence.
- Folds "catalogs should and provide a table's location" into the new third sentence, replacing the broken clause and adding the missing period.
- "is not a table-level concern" → "is outside the scope of this spec" — clearer about what the spec is leaving unspecified.

Author: This is also probably AI picking up prior revisions of the spec (it's using exact wording from earlier that was already rejected).

I fixed the grammatical error, the other suggestions are the opposite of what the goal of this section is.  We're not trying to say who provides the location, we also say the catalog should provide the location.

Reviewer: if location is not in the table metadata, who else can provide the location outside the catalog? is it correct to say that catalog `must` (instead of `should`) provide the location?

Author: No, there are other ways the location can be provided (e.g. a user can provide it directly when calling a `register_table` for example).  Additionally, there are engines that reference the metadata json path directly and then the user or engine needs to provide the path, which is why we don't make it a `MUST` requriement and we don't fully tie this to a catalog.

Writers are not allowed to commit files with a partition spec that contains a field with an unknown transform.

### Paths in Metadata

> Path strings stored in Iceberg metadata files are classified as one of two types:

Reviewer 2: "Path or location strings"?

If we changed that, then we could make the types "Absolute" and "Relative" to avoid the path/location issue a bit more.

Author: I'm generally referring to paths as the content stored within metadata location fields.  This allows us to continue to use standard conventions like "absolute paths" or "relative path", but the fields are location fields.

> * **Absolute path** -- A path string that includes a [URI scheme](https://datatracker.ietf.org/doc/html/rfc3986#section-3.1) (e.g., `s3:`, `gs:`, `hdfs:`, `file:`). Absolute paths are used as-is without modification.

Reviewer 13: My understanding (from https://datatracker.ietf.org/doc/html/rfc3986#section-3)
```
      URI         = scheme ":" hier-part [ "?" query ] [ "#" fragment ]
```
is that `s3`, `gs`, `hdfs`, `file` are examples of the scheme; the ':' is not part of the scheme.

Author: @wypoon you are correct that the `scheme` is just the preceding portion of the string, but all non-empty schemes must be followed by a `:`.  As these are examples of the prefix (not just scheme), I think including the colon is reasonable. I think it would be more confusing to exclude them as it would just be characters without a reference point, which the `:` provides.

We did remove the `//` after the colon because the authority is technically option and not required by the spec, but I feel like this should be fine.

Reviewer 13: I think that it is natural to read the parenthetical "e.g." next to the "URI scheme" link as giving examples of scheme. How about
```
* **Absolute path** -- A path string that includes a [URI scheme](https://datatracker.ietf.org/doc/html/rfc3986#section-3.1) (e.g., paths beginning with `s3:`, `gs:`, `hdfs:`, `file:`). Absolute paths are used as-is without modification.
```

> * **Absolute path** -- A path string that includes a [URI scheme](https://datatracker.ietf.org/doc/html/rfc3986#section-3.1) (e.g., `s3:`, `gs:`, `hdfs:`, `file:`). Absolute paths are used as-is without modification.

Reviewer 10: RFC 3986 defines an absolute URI as starting with a valid scheme followed by :, where a scheme is a letter followed by any combination of letters, digits, +, -, .. Should implementations validate against the full RFC, or can we define something simpler? For instance: a path is absolute if it contains `:` and no `/` appears before the first `:` This is sufficient for all real filesystem URIs and avoids more expensive scheme validation.

Reviewer 2: I think we should refer to the RFC as is done here.

Author: I agree that we should just refer to the RFC.  Implementations can make reasonable optimizations.  The intent here is to be clear about what the expected content is, not to force a full and rigorous validation anywhere a path is seen.  In practice, I we haven't really had any problems, but if there is a dispute about whether a representation is valid, we can always refer back to the RFC for clarification.

* **Relative path** -- A path string that does not include a URI scheme. Relative paths must be resolved against the table's base location before use.

> Prior to v4, all path fields must contain absolute paths. Starting with v4, path fields may contain either absolute or relative paths. Directory navigation symbols (`.` and `..`) and other file system conventions are not supported in relative paths.

Reviewer: > Directory navigation symbols (`.` and `..`) and other file system conventions are not supported in relative paths.

just for my knowledge, are they allowed in absolute paths?

Author: In some storage systems, yes.  For example in S3, those are just characters in the key space (no special meaning).  However, HadoopFileIO using a Hadoop Path will interpret it as directory navigation.

> Prior to v4, all path fields must contain absolute paths. Starting with v4, path fields may contain either absolute or relative paths. Directory navigation symbols (`.` and `..`) and other file system conventions are not supported in relative paths.

Reviewer 2: This isn't strictly true. I think it said "fully-qualified". Plus, we want to fix the case where local file paths (like `/path/without/scheme`) have leaked into metadata.

Author: I will update this to refer to v3- paths as 'fully-qualified', but we were not specific about that in the spec with the exception of the description for the `referenced_data_file` field.

Paths that start with `/` like your example may exist for local paths, but it's not limited to local paths and is very common for HDFS.  The configuration of the `core-site/hdfs-site` typically configures a default file system that includes the scheme and namenode address (e.g. `fs.defaultFS=hdfs://namenode-host:8020`).  All root based references start with `/` are then resolved based on the default fs value.

This means that there are many scenarios with HDFS where you will have these types of paths in existing metadata. Moving forward to v4, we would require these be canonicalized to `hdfs://...` (the default namenode can still be omitted so as to not to require encoding it in the path).  Cloud providers do not have this issue unless similarly configured for hdfs, but that's not typical.

> Prior to v4, all path fields must contain absolute paths. Starting with v4, path fields may contain either absolute or relative paths. Directory navigation symbols (`.` and `..`) and other file system conventions are not supported in relative paths.

Reviewer 2: ```suggestion
Prior to v4, all path fields must contain absolute paths. Starting with v4, path fields may contain either absolute or relative paths. Directory navigation symbols (`.` and `..`) and other file system conventions are not supported in any path or location string.
```

#### Path Resolution

Path resolution is the process of producing an absolute path from a relative path by combining it with the table's base location. If a path is absolute, it is used as-is. If a path is relative, it is concatenated with the table location to produce an absolute path:

> * If the path contains a URI scheme, it is absolute and is used without modification.

Reviewer 2: The paragraph above already says "If a path is absolute, it is used as-is", and we have a definition of "absolute" just above. I'd separate the last two sentences from that paragraph and combine them with these points to dedup.

* If the path does not contain a URI scheme, the resolved path is the table location followed by the relative path.

> Paths used as prefixes must not end in a path separator. The relative portion is appended to the prefix without introduction of any additional separator characters.

Reviewer: is `Paths used as prefixes` just `Table base location`? is latter more clear?

Author: It also applies to `write.metadata.path` for example during path construction.

> Paths used as prefixes must not end in a path separator. The relative portion is appended to the prefix without introduction of any additional separator characters.

Reviewer 13: For my understanding, can you please clarify:
Suppose the table location is "hdfs://ns/wh/foo.db/bar", and we are using relative paths; would a relative path for a data file look like "/data/00000-16-someuuid-0-00001.parquet", and the absolute path will then be "hdfs://ns/wh/foo.db/bar/data/00000-16-someuuid-0-00001.parquet"? Or should be the relative path for the data file be "data/00000-16-someuuid-0-00001.parquet" and we add the separator?
Further below, you discuss the `write.data.path`, which you state defaults to the value `data` if unspecified, and you write:
```
* If `write.data.path` is a relative path, the base is the table location followed by the `write.data.path` value.
```
If the table location must not end in a path separator, and we're using "data" for `write.data.path`, then we need to add the separator before the "data", don't we? Is "followed by" meant in a loose sense, or in a strict sense of being "appended ... without ... any additional separator characters"?

Author: We don't define or use separators in the path resolution on purpose (so that we don't have to define what the path separator is).  Some object storage doesn't require any path separator, so we want to remain agnostic about separators.

If you want a separator between the table location and the relative portion, then it should be prepended to the relative portion.  This holds true for `write.data.path` as well, so you would use `/data` if you want a path component relative to the table location.  Note, it also does not end in a path separator and the relative parts in the metadata should prefix a separator is one is expected.

> Paths used as prefixes must not end in a path separator. The relative portion is appended to the prefix without introduction of any additional separator characters.

Reviewer 10: I think this makes sense and would help with performance (since it's a simple concat instead of URL construction), but could really use an example. It is not super obvious that the relative paths needs to carry the leading /.

Reviewer 2: +1 for an example.

#### Path Relativization

Path relativization is the process of converting an absolute path to a relative path by removing the table location prefix. This is used when persisting paths to metadata files.

> * If an absolute path starts with the table location, the table location prefix should be removed and the remaining relative portion stored.

Reviewer 13: The table location must not end in a path separator; does the "remaining relative portion" include the beginning path separator or not?

Author: If you want a path separator (which is not required) then it should be at the front of the relative portion.

* If an absolute path does not start with the table location, it is stored as an absolute path.

#### Table Location Specification

> When the `location` field is present in table metadata, it is used directly as the table's base location. When the `location` field is not present (v4 and later), the table location must be provided.  How the table location is persisted/determined when not specified in metadata is outside the scope of the spec and is the responsibility of catalogs to track and provide.

Reviewer 2: ```suggestion
When the `location` field is present in table metadata, it is used directly as the table's base location. When the `location` field is not present (v4 and later), the table location must be provided. How the table location is persisted or determined when not specified in metadata is not a table-level concern; catalogs are intended to track and provide a table's location.
```

| _optional_ | _optional_ | _optional_ | _optional_ | **`snapshots`**             | A list of valid snapshots. Valid snapshots are snapshots for which all data files exist in the file system. A data file must not be deleted from the file system until the last snapshot in which it was listed is garbage collected.                                                                                                                                                                                                                                                                            |
| _optional_ | _optional_ | _optional_ | _optional_ | **`snapshot-log`**          | A list (optional) of timestamp and snapshot ID pairs that encodes changes to the current snapshot for the table. Each time the current-snapshot-id is changed, a new entry should be added with the last-updated-ms and the new current-snapshot-id. When snapshots are expired from the list of valid snapshots, all entries before a snapshot that has expired should be removed.                                                                                                                              |

> | _optional_ | _optional_ | _optional_ | _optional_ | **`metadata-log`**          | A list (optional) of timestamp and metadata file location pairs that encodes changes to the previous metadata files for the table. Each time a new metadata file is created, a new entry of the previous metadata file location should be added to the list. Tables can be configured to remove oldest metadata log entries and keep a fixed-size log of the most recent entries after a commit. In v4, metadata file locations may be relative paths resolved against the table location (see [Paths](#paths)). |

Reviewer 2: It's also a little surprising that this is updated to note relative/absolute handling, but other places in the spec aren't. I am fine with not updating each path/location, or linking to a section on file locations consistently. I would definitely have a section on file locations in the spec section and in that section discuss the difference between v3 and earlier ("must be fully-qualified") and v4 and later (must follow absolute/relative requirements).

| v1         | v2         | v3         | v4         | Field                       | Description                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                     |
| ---------- | ---------- |------------|------------|-----------------------------|-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|

> | _required_ | _required_ | _required_ | _required_ | **`format-version`**        | An integer version number for the format. Implementations must throw an exception if a table's version is higher than the supported version.                                                                                                                                                                                                                                                                                                                                                                    |

Reviewer 2: Minor: Our best practice for tables is to omit the last field's excess white space to avoid changes to every row like this.

Author: I think this was largely due to intellij formatted, but we added a new column so everything is getting pushed and it may be the diff show oddly.  I've tweaked it a bit so it should be consistent across the table's right-most bar.

    | _optional_ | _optional_ | _optional_ | **`partition-statistics`**  | A list (optional) of [partition statistics](#partition-statistics). |
    |            |            | _required_ | **`next-row-id`**           | A `long` higher than all assigned row IDs; the next snapshot’s `first-row-id`. See [Row Lineage](#row-lineage). |
    |            |            | _optional_ | **`encryption-keys`**       | A list (optional) of [encryption keys](#encryption-keys) used for table encryption. |

> === "v4"

Reviewer 2: I didn't expect this to add a new table for v4. Is this the same as the old table, except with a v4 requirement column?

Author: The purpose of adding the tabs was to separate V4 as we go to new structure.  Why would we update the old table and add tabs with not additional tabs if we're going to just remove that and add the tab later?

    | v4         | Field                       | Description |
    |------------|-----------------------------|-------------|
    | _required_ | **`format-version`**        | An integer version number for the format. Implementations must throw an exception if a table's version is higher than the supported version. |
    | _required_ | **`table-uuid`**            | A UUID that identifies the table, generated when the table is created. Implementations must throw an exception if a table's UUID does not match the expected UUID after refreshing metadata. |

>     | _optional_ | **`location`**              | The table's base location. This is used by writers to determine where to store data files, manifest files, and table metadata files. Must be an absolute path when present. See [Table Locations](#table-location-specification). |  

Reviewer 2: I'm assuming that this is the only line changed. Seems good to me.

    | _optional_ | _optional_ | _optional_ | **`partition-statistics`**  | A list (optional) of [partition statistics](#partition-statistics). |
    |            |            | _required_ | **`next-row-id`**           | A `long` higher than all assigned row IDs; the next snapshot’s `first-row-id`. See [Row Lineage](#row-lineage). |
    |            |            | _optional_ | **`encryption-keys`**       | A list (optional) of [encryption keys](#encryption-keys) used for table encryption. |
=== "v4"
    | v4         | Field                       | Description |
    |------------|-----------------------------|-------------|
    | _required_ | **`format-version`**        | An integer version number for the format. Implementations must throw an exception if a table's version is higher than the supported version. |
    | _required_ | **`table-uuid`**            | A UUID that identifies the table, generated when the table is created. Implementations must throw an exception if a table's UUID does not match the expected UUID after refreshing metadata. |

>     | _optional_ | **`location`**              | The table's base location. This is used by writers to determine where to store data files, manifest files, and table metadata files. Must be an absolute path when present. See [Table Locations](#table-location-specification). |  

Reviewer 11: When absent must be provided by the Catalog

Author: Same here, we don't want to restrict to catalogs only.

See comment here: https://github.com/apache/iceberg/pull/15630#discussion_r3251040825

    | _required_ | **`last-sequence-number`**  | The table's highest assigned sequence number, a monotonically increasing long that tracks the order of snapshots in a table. |
    | _required_ | **`last-updated-ms`**       | Timestamp in milliseconds from the unix epoch when the table was last updated. Each table metadata file should update this field just before writing. |
    | _required_ | **`last-column-id`**        | An integer; the highest assigned column ID for the table. This is used to ensure columns are always assigned an unused ID when evolving schemas. |

>     |            | **`schema`**                | The table’s current schema. (**Deprecated**: use `schemas` and `current-schema-id` instead) |

Reviewer 3: There is mixed usage of `table’s` and `table's`. Should we stick to one style?

Author: These fields are actually from the original table carried over.  I don't think we should update the wording here (it's not technically part of this spec change).

## Appendix E: Format version changes

### Version 4

Relative path support is added in v4.

Reading v3 metadata for v4:

* All location fields are treated as absolute paths

> * Any location field without a uri scheme prefix must prepend a scheme component consistent with v4

Reviewer: I thought absolute paths are required to start with a scheme component. why do we need this for pre-V4 metadata files?

Author: Someone in an earlier review commented that for local tables (or maybe DFS with default scheme?) may have entries that refer to `/<path>` as the file system root.  This just ensures that if you read from a v3 manifest, you produce valid paths for v4.

Reviewer: But it contradict with earlier sentences in this spec change. We said before V4 all file paths are absolute paths and absolute path is defined as including URI scheme.
```
 **Absolute path** -- A path string that includes a URI scheme and can be used directly.
...
Prior to v4, all path fields must contain absolute paths. 
```

Author: The definition of "Absolute Path" is not retroactive to v3.  It only applies to v4+.  All we're saying is in the rare event that you have v3- metadata, which requires all paths be absolute, that you make sure they are compliant with the v4 definition of Absolute Path.

Reviewer: > The definition of "Absolute Path" is not retroactive to v3. It only applies to v4+. 

When we define the term, we didn't say it only applies to V4+. I thought the term is retroactive to V3.

It sounds like the spec is both forbidding and accommodating the same thing. 

Maybe we can change the text and make the exception explicit and narrow: “Prior to v4, path fields are expected to be absolute. In practice, some legacy metadata may contain scheme-less root-relative paths such as /...; v4 readers may normalize those using an implementation-specific/default scheme.”

Or remove the migration bullet entirely if the behavior is too implementation-specific: the phrase “must prepend a scheme component consistent with v4” is underspecified and cause the tension on both forbidding and accommodating in this PR.

Author: I've updated the description in the spec to clarify that v3 paths are 'fully-qualified' which is the terminology we had for v3- and added minor clarification.  We never fully defined 'fully-qualified' but that's part of what we're trying to address here, so I wouldn't lean too hard into trying to now go back and rewrite the intent behind 'fully-qualified'

## Appendix E: Format version changes

### Version 4

> Relative path support is added in v4.

Reviewer: I feel we probably need to introduce sub headers for Version 4, as we have so many changes scoped for V4. if each feature has multiple paragraphs, it is hard to track.

The alternative is that each feature only has bullet points. E.g. the row lineage in V3 section below has a large bullet list.

I am slightly favor the former.

Reviewer 2: I think we want to focus on the content and not the formatting right now. It's great to have this section here to note changes, but I expect it to be updated quite a bit from how we handled this for the last 2 versions.

Reading v3 metadata for v4:

* All location fields are fully-qualified paths and interpreted as absolute paths for v4
* Any location field without a uri scheme prefix must prepend a scheme component consistent with v4 absolute paths

Writing v4 metadata:

* Table metadata JSON:
    * `location` is now optional and must be absolute when present
    * When not present, the table location must be managed externally and provided when loading the metadata
* Location fields in all metadata structures may contain relative paths
* Writers should produce relative paths by default for files that reside under the table location

> * Absolute paths must be used for files that do not share a common prefix with the table location

Reviewer: Same "shares a common prefix" ambiguity flagged in the inline at line 1905. Line 1750 above ("files that reside under the table location") is also informal terminology for what is now formally defined in [Path Relativization](#path-relativization).

Suggested rewording for lines 1750–1751:

> * Writers should produce relative paths by default for files whose absolute path starts with the table location followed by the URI separator `/`.
> * Absolute paths must be used for all other files.

This way the writer rule is exactly the Path Relativization rule applied — no chance of producing a relative form that doesn't round-trip safely under table relocation.

Author: This is just a summary of changes not the specification itself.  While I think it made sense to add in the spec section, adding it here isn't really helpful and just makes it more complicated.

Reviewer: > Absolute paths must be used for all other files.

line 1751 can be simplified as `for all other files`. We don't even need to mention `that don't share a common prefix ...`, which can leads to misinterpretation.

But I will leave it up to you.

Author: That would make the last bullet dependent upon the prior bullet, which while technically correct if you read them all together, I think this is more clear because each bullet stands as a complete statement on its own.

## Appendix E: Format version changes

### Version 4

Relative path support is added in v4.

> Reading v3 metadata for v4:

Reviewer 7: nit: v3 or prior metadata for v4

* All location fields are fully-qualified paths and interpreted as absolute paths for v4

> * Any location field without a uri scheme prefix must prepend a scheme component consistent with v4 absolute paths

Reviewer 4: nit: `URI` to be consistent

Writing v4 metadata:

* Table metadata JSON:
    * `location` is now optional and must be absolute when present

>     * When not present, the table location must be managed externally and provided when loading the metadata

Reviewer 11: should we just say catalog here? Or do we want to leave this open?

Author: Same situation here.  We want to leave this open as it's not catalog only.

https://github.com/apache/iceberg/pull/15630#discussion_r3251040825

This section covers topics not required by the specification but recommendations for systems implementing the Iceberg specification to help maintain a uniform experience.

### Path Construction

Path construction is the process by which new file locations are created for output files referenced by metadata. While the specific construction logic is not strictly required by the spec, the following guidance is provided for reference implementations to encourage consistency.

> The table properties `write.metadata.path` and `write.data.path` control where metadata and data files are written relative to the table location. When not specified, these default to the values `metadata` and `data` respectively.

Reviewer 13: I don't believe that `write.metadata.path` and `write.data.path` are mentioned in the spec before this. However, my understanding of the table properties according to docs/docs/configuration.md
```
| write.data.path                                     | table location + /data      | Base location for data files                                                                                                                                                                                                                       |
| write.metadata.path                                 | table location + /metadata  | Base location for metadata files                                                                                                                                                                                                                   |
```
is that `write.metadata.path` and `write.data.path` default to table location + /data and table location + /metadata respectively.
In other words, since relative paths are not supported up to now, the default values of `write.metadata.path` and `write.data.path` have to be the absolute paths given by the above.

I believe that we now want to allow `write.metadata.path` and `write.data.path` to be relative paths, and want the defaults to be `metadata` and `data` respectively. Would you agree?

I think it'd be more accurate to write:

"The table properties `write.metadata.path` and `write.data.path` control where metadata and data files are written. When not specified, these default respectively to `metadata` and `data` folders under the table location. As we now allow paths to be relative, the default values for the properties would be `metadata` and `data` respectively."

I know that you're trying to avoid words like folder and directory, but I couldn't think of how to express the idea of something "under a table location" otherwise.

Author: This addition is under the appendix as implementation notes.  Nothing in this section is defines a spec requirement.  It's mostly just about how the reference implementation (and many that follow it) behaves for consistency in layout.

The lines below describe what you're trying to explain with relative or non-relative properties.  We do avoid folder/directory wording, but description below covers this.

This section covers topics not required by the specification but recommendations for systems implementing the Iceberg specification to help maintain a uniform experience.

### Path Construction

Path construction is the process by which new file locations are created for output files referenced by metadata. While the specific construction logic is not strictly required by the spec, the following guidance is provided for reference implementations to encourage consistency.

> The table properties `write.metadata.path` and `write.data.path` control where metadata and data files are written relative to the table location. When not specified, these default to the values `metadata` and `data` respectively.

Reviewer: This sentence contradicts the bullets directly below. The opening says the properties control where files are written **"relative to the table location,"** but the bullets at lines 1897 and 1902 explicitly handle the case where `write.metadata.path` / `write.data.path` is an **absolute** path that's used directly, bypassing the table location.

Suggested rewrite:

> The table properties `write.metadata.path` and `write.data.path` control where metadata and data files are written. Each property may be an absolute path or a relative path; relative paths are resolved against the table location. When not specified, these default to the relative values `metadata` and `data`.

This removes the contradiction and surfaces the absolute/relative distinction upfront so the bullets that follow read as enumeration rather than exceptions.

Author: I've remove the "relative to the table location" as it's addressed in the bulleted references.

For all metadata files:

* If `write.metadata.path` is an absolute path, it is used directly as the base for new metadata files.
* If `write.metadata.path` is a relative path, the metadata base is the table location joined to the `write.metadata.path` value with a URI separator `/`.

For data files:

* If `write.data.path` is an absolute path, it is used directly as the base for new data files.
* If `write.data.path` is a relative path, the base is the table location joined to the `write.data.path` value with a URI separator `/`.

> When persisting paths into metadata, writers should relativize paths against the table location (see [Path Relativization](#path-relativization)). If a file's absolute path shares a common prefix with the table location, the relative portion should be stored. Otherwise, the absolute path should be stored.

Reviewer: This recommendation wasn't updated to match the new boundary-aware rule in [Path Relativization](#path-relativization). "Shares a common prefix" is the original ambiguous phrasing — the same one that allows `s3://bucket/table_v2/file` to be "under" `s3://bucket/table` under a byte-prefix interpretation.

Suggested rewording to mirror the normative section:

> When persisting paths into metadata, writers should relativize paths against the table location (see [Path Relativization](#path-relativization)). If a file's absolute path starts with the table location followed by the URI separator `/`, the relative portion (after removing the prefix and separator) should be stored. Otherwise, the absolute path should be stored.

Author: Fixed

> When persisting paths into metadata, writers should relativize paths against the table location (see [Path Relativization](#path-relativization)). If a file's absolute path shares a common prefix with the table location, the relative portion should be stored. Otherwise, the absolute path should be stored.

Reviewer 2: If allowed by the table version?
